finality-aleph/src/party.rs

```diff
@@ -70,194 +101,276 @@ fn get_node_index(authorities: &[AuthorityId], my_id: &AuthorityId) -> Option<No
     phantom: PhantomData<BE>,
-    finalization_proposals_rx: mpsc::UnboundedReceiver<JustificationNotification<B>>,
     metrics: Option<Metrics<B::Header>>,
+    authority_justification_tx: mpsc::UnboundedSender<JustificationNotification<B>>,
 }
 
-/// If we are on the authority list for the given session, runs an
-/// AlephBFT task and returns `true` upon completion. Otherwise, immediately returns `false`.
-async fn maybe_run_session_as_authority<B, C, BE, SC>(
+async fn run_aggregator<B, C, BE>(
+    rmc_network: RmcNetwork<B>,
+    multikeychain: MultiKeychain,
+    mut ordered_batch_rx: mpsc::UnboundedReceiver<OrderedBatch<B::Hash>>,
+    justification_tx: mpsc::UnboundedSender<JustificationNotification<B>>,
+    client: Arc<C>,
+    session: Session<AuthorityId, NumberFor<B>>,
+    mut exit_rx: futures::channel::oneshot::Receiver<()>,
+) where
+    B: Block,
+    C: crate::ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: aleph_primitives::AlephSessionApi<B, AuthorityId, NumberFor<B>>,
+    BE: Backend<B> + 'static,
+{
+    let mut aggregator = BlockSignatureAggregator::new(rmc_network, &multikeychain);
+    let mut last_finalized = client.info().finalized_hash;
+    let mut last_block_seen = false;
+    let current_stop_h = session.stop_h;
+    loop {
+        tokio::select! {
+            maybe_batch = ordered_batch_rx.next() => {
+                if let Some(batch) = maybe_batch {
+                    trace!(target: "afa", "Received batch {:?} in aggregator.", batch);
+                    if last_block_seen {
+                        //This is only for optimization purposes.
+                        continue;
+                    }
+                    for new_hash in batch {
+                        let to_finalize_headers = chain_extension_step(last_finalized, new_hash, client.as_ref());
+                        for header in to_finalize_headers.iter() {
+                            if *header.number() <= current_stop_h {
+                                aggregator.start_aggregation(header.hash()).await;
+                                last_finalized = header.hash();
+                            }
+                            if *header.number() >= current_stop_h {
+                                aggregator.notify_last_hash();
+                                last_block_seen = true;
+                                break;
+                            }
+                        }
+                    }
+                } else {
+                    debug!(target: "afa", "Batches ended in aggregator. Terminating.");
+                    return;
+                }
+            }
+            multisigned_hash = aggregator.next_multisigned_hash() => {
+                if let Some((hash, multisignature)) = multisigned_hash {
+                    let number = client.number(hash).unwrap().unwrap();
+                    // The unwrap might actually fail if data availability is not implemented correctly.
```

Comment:
And it isn't right now, right? Not a blocker, just making sure I understand.

Although technically against the Zero Bug Policy, but I'm more and more convinced that one is just pure fiction...

Reply:
Right now it is not at all implemented, Maciek working on it.